tools/python/util/onnx_model_utils.py

```diff
@@ -0,0 +1,311 @@
+    :param graph: Graph to update
+    :param input_name: Name of graph input to update.
+    :param fixed_shape: Shape to use.
+    '''
+
+    for i in graph.input:
+        if i.name == input_name:
+            if not i.type.HasField("tensor_type"):
+                raise ValueError(f'Input {input_name} is not a tensor')
+
+            # graph input are required to have a shape to provide the rank
+            shape = i.type.tensor_type.shape
+            if len(shape.dim) != len(fixed_shape):
+                raise ValueError(
+                    f'Rank mismatch. Existing:{len(shape.dim)} Replacement:{len(fixed_shape)}')
+
+            idx = 0
+            for dim in shape.dim:
+                # check any existing fixed dims match
+                if dim.HasField('dim_value'):
+                    if dim.dim_value != fixed_shape[idx]:
+                        raise ValueError(
+                            f"Can't replace existing fixed size of {dim.dim_value} with {fixed_shape[idx]} "
+                            f"for dimension {idx + 1}")
+                elif dim.HasField('dim_param'):
+                    # replacing a dim_param so have to do that through the entire graph
+                    make_dim_param_fixed(graph, dim.dim_param, fixed_shape[idx])
+                else:
+                    # replacing an unknown dim
+                    dim.Clear()
+                    dim.dim_value = fixed_shape[idx]
+
+                idx += 1
+
+            return
+
+    raise ValueError(f'Input {input_name} was not found in graph inputs. '
+                     f'Valid input names are: {",".join([i.name for i in graph.input])}')
+
+
+def _create_producer_consumer_link(node_to_producers: dict, node_to_consumers: dict,
+                                   producer: onnx.NodeProto, consumer: onnx.NodeProto):
+    '''
+    Create links between two nodes for a value produced by one and consumed by the other.
+    :param node_to_producers: Map of NodeProto to set of nodes that produce values the node consumes as inputs.
+    :param node_to_consumers: Map of NodeProto to set of nodes that consume values the node produces as outputs.
+    :param producer: Producer node
+    :param consumer: Consumer node
+    '''
+
+    if consumer not in node_to_producers:
+        node_to_producers[consumer] = set()
+
+    if producer not in node_to_consumers:
+        node_to_consumers[producer] = set()
+
+    # add entry mapping this node to the producer of this input
+    node_to_producers[consumer].add(producer)
+    node_to_consumers[producer].add(consumer)
+
+
+def _map_node_dependencies(graph: onnx.GraphProto, node_to_producers: dict, node_to_consumers: dict):
+    graph_inputs = set([i.name for i in graph.input])
+    initializers = set([i.name for i in graph.initializer])
+
+    # map of value name to node that creates it. copy parent values but override if values get shadowed
```

Comment:
>copy parent values but override if values get shadowed [](http://example.com/codeflow?start=49&length=54)

didn't quite understand this. looks like it only gets set from node outputs, when is overriding possible? at most one node produces a given value right?

Reply:
oh is the shadowing of implicit inputs from the parent graph?

Reply:
a value in the subgraph will shadow any value of the same name in the parent graph. 

that also means there should never be an implicit input for a value who's scope is the subgraph, as the implicit input is simply providing access to the value in the outer scope (and that value is being shadowed).